L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+    tkMuons.push_back(l1tkmu);
+  }
+
+  return;
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void L1TkMuonProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
+
+L1TkMuonProducer::PropState L1TkMuonProducer::propagateToGMT(const L1TkMuonProducer::L1TTTrackType& tk) const {
+  auto p3 = tk.momentum();
+  float tk_pt = p3.perp();
+  float tk_p = p3.mag();
+  float tk_eta = p3.eta();
+  float tk_aeta = std::abs(tk_eta);
+  float tk_phi = p3.phi();
+  float tk_q = tk.rInv() > 0 ? 1. : -1.;
+  float tk_z = tk.POCA().z();
+  if (!correctGMTPropForTkZ_)
+    tk_z = 0;
+
+  L1TkMuonProducer::PropState dest;
+  if (tk_p < 3.5)
+    return dest;
+  if (tk_aeta < 1.1 && tk_pt < 3.5)
+    return dest;
+  if (tk_aeta > 2.5)
+    return dest;
+
+  //0th order:
+  dest.valid = true;
+
+  float dzCorrPhi = 1.;
+  float deta = 0;
+  float etaProp = tk_aeta;
+
+  if (tk_aeta < 1.1) {
+    etaProp = 1.1;
+    deta = tk_z / 550. / cosh(tk_aeta);
+  } else {
+    float delta = tk_z / 850.;  //roughly scales as distance to 2nd station
+    if (tk_eta > 0)
+      delta *= -1;
+    dzCorrPhi = 1. + delta;
+
+    float zOzs = tk_z / 850.;
+    if (tk_eta > 0)
+      deta = zOzs / (1. - zOzs);
+    else
+      deta = zOzs / (1. + zOzs);
+    deta = deta * tanh(tk_eta);
+  }
+  float resPhi = tk_phi - 1.464 * tk_q * cosh(1.7) / cosh(etaProp) / tk_pt * dzCorrPhi - M_PI / 144.;
+  if (resPhi > M_PI)
+    resPhi -= 2. * M_PI;
+  if (resPhi < -M_PI)
+    resPhi += 2. * M_PI;
+
+  dest.eta = tk_eta + deta;
+  dest.phi = resPhi;
+  dest.pt = tk_pt;  //not corrected for eloss
+
+  dest.sigmaEta = 0.100 / tk_pt;  //multiple scattering term
+  dest.sigmaPhi = 0.106 / tk_pt;  //need a better estimate for these
+  return dest;
+}
+
+double L1TkMuonProducer::sigmaEtaTP(const RegionalMuonCand& l1mu) const {
+  float l1mu_eta = l1mu.hwEta() * 0.010875;
+  if (std::abs(l1mu_eta) <= 1.55)
+    return 0.0288;
+  else if (std::abs(l1mu_eta) > 1.55 && std::abs(l1mu_eta) <= 1.65)
+    return 0.025;
+  else if (std::abs(l1mu_eta) > 1.65 && std::abs(l1mu_eta) <= 2.4)
+    return 0.0144;
+  return 0.0288;
+}
+
+double L1TkMuonProducer::sigmaPhiTP(const RegionalMuonCand& mu) const { return 0.0126; }
+
+// ------------------------------------------------------------------------------------------------------------
+
+std::vector<L1TkMuMantraDF::track_df> L1TkMuonProducer::product_to_trkvec(const L1TTTrackCollectionType& l1tks) {
+  std::vector<L1TkMuMantraDF::track_df> result(l1tks.size());
+  for (uint itrk = 0; itrk < l1tks.size(); ++itrk) {
+    auto& trk = l1tks.at(itrk);
+
+    result.at(itrk).pt = trk.momentum().perp();
```

Comment:
this would be a lot easier if you gave the struct a constructor, then you could just use emplace

Reply:
Agree.  But can do in the next version.